_drafts/restless-restfull.md

```diff
@@ -0,0 +1,456 @@
+
+In a distributed system making idempotent operations explicit is
+important for failure recovery. If an operation is idempotent
+and its answer timeouts (perhaps a node went down, who knows ?)
+you can safely retransmit it, if the previous one was actually
+executed this will generate no problem's at all since the operation
+is idempotent. If this does not seem that important to you yet
+checkout this
+[awesome post on cockroachdb passing the jepsen test](https://www.cockroachlabs.com/blog/cockroachdb-beta-passes-jepsen-testing/).
+
+Quoting:
+
+```
+The sets test (described in section 2.7) revealed a bug
+when it was refactored to use a single auto-committed INSERT
+statement instead of a multi-statement BEGIN/INSERT/COMMIT transaction
+(using a table with no primary key).
+
+Single statements can be retried by the server after certain
+errors (multi-statement transactions generally cannot,
+because the results of previous statements have already been
+returned to the client).
+
+In this case, the statement was attempted once, timed
+out due to a network failure, then retried,
+and both attempts eventually succeeded
+(this was possible because the table had an
+auto-generated primary key, so the two insertion
+attempts were not using the same PK values).
+```
+
+Distributed systems are hard, being sure of what can be retried and what
+can't is essential. So right now we have 3 categories of operations:
+
+* Side Effect Free
+* Idempotent (with side effects)
+* Non-Idempotent  (with side effects)
+
+To give a more simple example, deleting a resource using the **DELETE**
+method is safe for retransmission since deleting something twice
+ends up on the same final state, that thing has been deleted.
+
+It is important to not mix leaving the system at the same state
+and giving the same answer. The first delete will return OK and the second
+one will fail, the operation is still idempotent because the final state
+of the system is the same (the resource has been deleted).
+
+Another interesting property of operations is commutativity, since they
+enable parallelization, but AFAIK there is nothing on REST about that.
+
+We saw some examples of side effects free operations and idempotent ones,
+this leaves us with the non-idempotent one, the beloved **POST** method.
+
+What should be the semantics of **POST** ? This is one of the biggest
+sources of bike shedding on RESTfullness because of the addiction
+to creating resources, that does not map well with some sorts of processes
+(yeah, REST can not map well to stuff,
+it is real, it happens, more on that later).
+
+An example from a RESTful guideline:
+
+```
+POST requests are idiomatically used to create
+single resources on a collection resource endpoint,
+```
+
+Which leads to:
+
+```
+POST request should only be applied to collection resources,
+and normally not on single resource, as this has an undefined semantic
+```
+
+TODO: Again we go back to thinking only on CRUD stuff, collections and
```

Comment:
yeah, I would like to read your opinion regarding other kinds of application, like a search app, a bot API, consumers, and so on. :-)

Reply:
I dont have too much experience with search API's, but I got some with the captcha/proxy API's. Specially a captcha API is something that we are working right now. I'll try to ellaborate them in more detail soon =D